src/plugins/discord/mirror.test.js

```diff
@@ -0,0 +1,338 @@
+// @flow
+
+import Database from "better-sqlite3";
+import {Mirror} from "./mirror";
+import dedent from "../../util/dedent";
+import {
+  type Channel,
+  type Message,
+  type GuildMember,
+  type Snowflake,
+  type Emoji,
+} from "./models";
+
+describe("plugins/discord/mirror", () => {
+  const customEmoji = (): Emoji => ({id: "id", name: "name"});
+  const genericEmoji = (): Emoji => ({id: null, name: "🐙"});
+
+  const testMessage = (id: Snowflake): Message => ({
+    id: id,
+    channelId: "2",
+    authorId: "3",
+    timestampMs: Date.parse("2020-03-03T23:35:10.615000+00:00"),
+    content: "Just going to drop this here",
+    reactionEmoji: [customEmoji()],
+    nonUserAuthor: false,
+    mentions: ["1", "23"],
+  });
+
+  const testMember = (userId: Snowflake): GuildMember => ({
+    user: {
+      id: userId,
+      username: "username",
+      discriminator: "disc",
+      bot: true,
+    },
+    nick: "nickname",
+  });
+
+  describe("constructor", () => {
+    it("initializes a new database succsessfully", () => {
+      const db = new Database(":memory:");
+      expect(() => new Mirror(db, "0")).not.toThrow();
+    });
+
+    it("rejects a different config", () => {
+      const db = new Database(":memory:");
+      const _ = new Mirror(db, "0");
+      expect(() => new Mirror(db, "1")).toThrow(
+        "Database already populated with incompatible server or version"
+      );
+    });
+
+    it("creates the right set of tables", () => {
+      const db = new Database(":memory:");
+      new Mirror(db, "0");
+      const tables = db
+        .prepare("SELECT name FROM sqlite_master WHERE type = 'table'")
+        .pluck()
+        .all();
+      expect(tables.sort()).toEqual(
+        Array.from(
+          new Set([
+            "meta",
+            "channels",
+            "members",
+            "messages",
+            "message_reactions",
+            "message_mentions",
+          ])
+        ).sort()
+      );
+    });
+  });
+
+  describe("members", () => {
+    it("inserts members", async () => {
+      const member = testMember("1");
+      const db = new Database(":memory:");
+      const mirror = new Mirror(db, "0");
+      mirror.addMember(member);
+      const result = db
+        .prepare(
+          dedent`\
+        SELECT
+          user_id,
+          username,
+          discriminator,
+          bot,
+          nick
+        FROM members`
+        )
+        .get();
```

Comment:
Nit: indentation here is strange, can you ensure the SQL is indented further than the `.prepare` call?

Reply:
updated